File: RxAdapter/src/main/java/com/kronos/rxadapter/RequestAdapter.java
package com.kronos.rxadapter;

import com.kronos.volley.RequestResponse;
import com.kronos.volley.VolleyError;
import com.kronos.volley.toolbox.NetResponse;
import com.kronos.volley.toolbox.StringRequest;

/**
 * Created by Leif Zhang on 2017/4/1.
 * Email leifzhanggithub@gmail.com
 */

public class RequestAdapter implements RequestResponse.Listener<NetResponse>, RequestResponse.ErrorListener {
    private NetResponse netResponse;
    private boolean isReady = false;
    private boolean isFinish = false;
    private VolleyError volleyError;
    private StringRequest request;

    public boolean isReady() {
        boolean ready = isReady;
        if (isReady) {
            isReady = false;
        }
        return ready;
    }

    public NetResponse getNetResponse() throws VolleyError {
        if (volleyError != null) {
            throw volleyError;
        }
        return netResponse;
    }

    public boolean isFinish() {
        return isFinish;

    }

    public RequestAdapter(final StringRequest request) {
        this.request = request;
        request.setRequestListener(this).setErrorListener(this);
    }

    @Override
    public void onErrorResponse(VolleyError error) {
        volleyError = error;
        isFinish = true;
        isReady = true;
    }

    @Override
    public void onResponse(NetResponse response) {
        try {
            netResponse = response;
            isReady = true;
            isFinish = !request.isRefreshNeed() || !response.isCache;
        } catch (Exception e) {
            e.printStackTrace();
            onErrorResponse(new VolleyError());
        }
    }
}

File: RxAdapter/src/main/java/com/kronos/rxadapter/RequestAdapter.kt
package com.kronos.rxadapter

import com.kronos.volley.Response
import com.kronos.volley.VolleyError
import com.kronos.volley.toolbox.NetResponse
import com.kronos.volley.toolbox.StringRequest

import rx.Observable
import rx.Subscriber

/**
 * Created by Leif Zhang on 2017/4/1.
 * Email leifzhanggithub@gmail.com
 */

class RequestAdapter(private val request: StringRequest) : Response.Listener<NetResponse>, Response.ErrorListener, Observable.OnSubscribe<NetResponse> {
    private var netResponse: NetResponse? = null
    private var isReady = false
    private var isFinish = false
    private var volleyError: VolleyError? = null
    private var subscriber: Subscriber<in NetResponse>? = null

    init {
        request.setRequestListener(this).errorListener = this
    }

    override fun onErrorResponse(error: VolleyError) {
        volleyError = error
        isFinish = true
        isReady = true
        subscriber!!.onError(volleyError)
    }

    override fun onResponse(response: NetResponse) {
        try {
            netResponse = response
            isReady = true
            isFinish = !request.isRefreshNeed || !response.isCache
            subscriber!!.onNext(netResponse)
            if (isFinish) {
                subscriber!!.onCompleted()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            subscriber?.onError(e)
        }

    }

    override fun call(subscriber: Subscriber<in NetResponse>) {
        this.subscriber = subscriber
    }
}
